### DISCUSSION TITLE
Use Buffer in custom client

### DISCUSSION THREAD

**Discussion by ylkuo** [mozilla/hubs#3631] [Hubs Cloud Questions] [answered]
I use a package that requires Buffer. I turned on this line in webpack for the local demo and it works well:
https://github.com/mozilla/hubs/blob/36d8941ce261fbb89228cca71ce3c8f2e7d52324/webpack.config.js#L245
Is there any way/workaround to use Buffer for the custom client deployed to aws?

**Comment by robinkwilson** [comment]
Run `npm run deploy` to deploy your custom client. Webpack and the deploy script will bundle that new dependency into your custom client.

If that doesn't work, take a peek at deploy.js script.

**Reply by ylkuo** [reply]
Thank you for the suggestion but we got some errors when running `npm run deploy` after turning `Buffer` to `true`. It was an `EPERM` error as follows:

```
> node -r @babel/register -r esm -r ./scripts/shim scripts/deploy.js

Deploying to xxx.com.
  / Building Client.(node:8160) UnhandledPromiseRejectionWarning: Error: Command failed: npm ci
npm WARN prepare removing existing node_modules/ before installation
npm ERR! code EPERM
npm ERR! syscall unlink
npm ERR! path D:\code\node_modules\utf-8-validate\prebuilds\win32-x64\node.napi.node
npm ERR! errno -4048
npm ERR! Error: EPERM: operation not permitted, unlink 'D:\code\node_modules\utf-8-validate\prebuilds\win32-x64\node.napi.node'
npm ERR!  [OperationalError: EPERM: operation not permitted, unlink 'D:\code\node_modules\utf-8-validate\prebuilds\win32-x64\node.napi.node'] {
npm ERR!   cause: [Error: EPERM: operation not permitted, unlink 'D:\code\node_modules\utf-8-validate\prebuilds\win32-x64\node.napi.node'] {
npm ERR!     errno: -4048,
npm ERR!     code: 'EPERM',
npm ERR!     syscall: 'unlink',
npm ERR!     path: 'D:\\code\\node_modules\\utf-8-validate\\prebuilds\\win32-x64\\node.napi.node'
npm ERR!   },
npm ERR!   isOperational: true,
npm ERR!   errno: -4048,
npm ERR!   code: 'EPERM',
npm ERR!   syscall: 'unlink',
npm ERR!   path: 'D:\\code\\node_modules\\utf-8-validate\\prebuilds\\win32-x64\\node.napi.node'
npm ERR! }
```

We tried to remove the `node_modules` or run `npm run deploy` as administrators but they all show the same error. Any idea to fix this or where do we need to look into in the deploy.js script? Thank you again!

**Reply by robinkwilson** [reply]
Does `npm ci` complete when you run it in the project root with `buffer: true`?

Can you run `npm -v` and `node -v` in the command line and tell me what versions get printed?

**Reply by ylkuo** [reply]
Yes, `npm ci` completed.
We have `v14.15.0` for `node -v` and `6.14.8` for `npm -v`.
Thank you!

**Reply by RathodDeven** [reply]
Hi, @ylkuo did you found any solution to deploy on custom client on aws.. ? I ran into same problem was able to run the libraries that required buffer: true in local but could not deploy on custom client.. Would love to here if you have found any solution or workaround ..

**Reply by ylkuo** [reply]
@RathodDeven No, we ended up using an implementation without setting `buffer: true`

**Comment by robinkwilson** [comment, marked as answer]
Although I'm not 100% sure how `Buffer=true` impacts sending the bundled client code to S3, the deploy script/deploy system isn't configured for it currently.

You're welcome to take a look at the deploy.js script (runs with `npm run deploy`), if you figure out a way to get S3 to work with `Buffer=true`, I'd be overjoyed to see. I'm sure everyone agrees, we'd LOVE the deploy process to be quicker. 

You could make an issue/feature request in the hubs-cloud repo so it's seen by the team! -- I just can't guarantee it will be worked on anytime soon unfortunately.
